NAPS2.Lib/Dependencies/DownloadController.cs

```diff
@@ -97,80 +53,110 @@ public void Stop()
-        }
-        else
-        {
-            _urlIndex = 0;
-        }
-        if (FilesDownloaded > 0 && _urlIndex == 0)
-        {
-            var prev = _filesToDownload[FilesDownloaded - 1];
-            var filePath = Path.Combine(prev.TempFolder!, prev.DownloadInfo.FileName);
+            if (result == null)
+            {
+                DownloadComplete?.Invoke(this, EventArgs.Empty);
+                DownloadError?.Invoke(this, EventArgs.Empty);
+                return false;
+            }
+
+            string tempFolder = Path.Combine(_scanningContext.TempFolderPath, Path.GetRandomFileName());
+            Directory.CreateDirectory(tempFolder);
+            string p = Path.Combine(tempFolder, fileToDownload.DownloadInfo.FileName);
             try
             {
-                var preparedFilePath = prev.DownloadInfo.Format.Prepare(filePath);
-                prev.FileCallback(preparedFilePath);
+                result.Position = 0;
+                var preparedFilePath = fileToDownload.DownloadInfo.Format.Prepare(result, p);
+                fileToDownload.Install(preparedFilePath);
             }
             catch (Exception ex)
             {
                 _logger.LogError(ex, "Error preparing downloaded file");
                 DownloadError?.Invoke(this, EventArgs.Empty);
             }
-            Directory.Delete(prev.TempFolder!, true);
-        }
-        if (FilesDownloaded >= _filesToDownload.Count)
-        {
-            DownloadComplete?.Invoke(this, EventArgs.Empty);
-            _completionSource.SetResult(true);
-            return;
-        }
-        if (_urlIndex >= _filesToDownload[FilesDownloaded].DownloadInfo.Urls.Count)
-        {
-            DownloadComplete?.Invoke(this, EventArgs.Empty);
-            _completionSource.SetResult(false);
-            DownloadError?.Invoke(this, EventArgs.Empty);
-            return;
+            FilesDownloaded++;
+            Directory.Delete(tempFolder, true);
         }
-        var next = _filesToDownload[FilesDownloaded];
-        next.TempFolder = Path.Combine(_scanningContext.TempFolderPath, Path.GetRandomFileName());
-        Directory.CreateDirectory(next.TempFolder);
-        _client.DownloadFileAsync(new Uri(next.DownloadInfo.Urls[_urlIndex]), Path.Combine(next.TempFolder, next.DownloadInfo.FileName));
+
+        DownloadComplete?.Invoke(this, EventArgs.Empty);
+        return true;
     }
 
-    private string CalculateSha1(string filePath)
+    private string CalculateSha1(Stream stream)
     {
         using var sha = SHA1.Create();
-        using FileStream stream = File.OpenRead(filePath);
         byte[] checksum = sha.ComputeHash(stream);
         string str = BitConverter.ToString(checksum).Replace("-", String.Empty).ToLowerInvariant();
         return str;
     }
 
-    private class QueueItem
+    private record class CustomDownloadHandler(DownloadInfo DownloadInfo, Action<string> FileCallback) : IExternalComponent
```

Comment:
No we don't need to use `IExternalComponent`. `QueueItem` is getting very small because `TempFolder` is not necessary anymore.

If you want me to keep using `QueueItem` let me know.

Reply:
Yeah, let's just do a QueueItem record class.